cmd/downloader/downloader.go

```diff
@@ -218,74 +105,57 @@ func ExtractZipArchive(archiveFile, extractPath string, service Service) {
 	archive, err := gzip.NewReader(file)
-	CheckError(err)
+	utils.CheckError(err)
 	tarReader := tar.NewReader(archive)
+	if len(service.ArchivePath) > 0 {
+		outputPath = filepath.Join(extractPath, service.ArchivePath)
+	}
+	if len(service.OutputPath) > 0 {
+		outputPath = filepath.Join(extractPath, service.OutputPath)
+	}
 	for {
 		header, err := tarReader.Next()
 		if err == io.EOF {
 			break
 		}
-		CheckError(err)
+		utils.CheckError(err)
 		if strings.HasSuffix(header.Name, service.ArchivePath) {
-			var path string
-			if len(service.OutputPath) > 0 {
-				path = filepath.Join(extractPath, service.OutputPath)
-			} else if len(service.ArchivePath) > 0 {
-				path = filepath.Join(extractPath, service.ArchivePath)
-			} else {
-				path = filepath.Join(extractPath, header.Name)
+			if outputPath == "" {
+				outputPath = filepath.Join(extractPath, header.Name)
 			}
-			glog.Infof("Extracting to %q", path)
-			outfile, err := os.Create(path)
-			CheckError(err)
+			glog.Infof("Extracting to %q", outputPath)
+			outfile, err := os.Create(outputPath)
+			utils.CheckError(err)
 			if _, err := io.Copy(outfile, tarReader); err != nil {
-				glog.Errorf("Failed to create file: %q", path)
+				glog.Errorf("Failed to create file: %q", outputPath)
 			}
 			outfile.Chmod(fs.FileMode(header.Mode))
 			outfile.Close()
 		}
 	}
+	return nil
 }
 
-func Run(buildFlags BuildFlags) {
-	cliFlags := CliFlags{}
-	flag.Set("logtostderr", "true")
-	vFlag := flag.Lookup("v")
-	fs := flag.NewFlagSet(constants.APP_NAME, flag.ExitOnError)
-
-	fs.StringVar(&cliFlags.Verbosity, "v", "", "Log verbosity.  {4|5|6}")
-	fs.StringVar(&cliFlags.Platform, "platform", runtime.GOOS, "One of linux/windows/darwin")
-	fs.StringVar(&cliFlags.Architecture, "architecture", runtime.GOARCH, "System architecture (amd64/arm64)")
-	fs.StringVar(&cliFlags.DownloadPath, "path", fmt.Sprintf(".%sbin", string(os.PathSeparator)), "Path to store binaries")
-	fs.StringVar(&cliFlags.ManifestFile, "manifest", "manifest.yaml", "Path to manifest yaml file")
-	fs.BoolVar(&cliFlags.SkipDownloaded, "skip-downloaded", false, "Skip already downloaded archive (if found)")
-	fs.BoolVar(&cliFlags.Cleanup, "cleanup", true, "Cleanup downloaded archives after extraction")
-
-	ff.Parse(
-		fs, os.Args[1:],
-		ff.WithConfigFileFlag("config"),
-		ff.WithConfigFileParser(ff.PlainParser),
-		ff.WithEnvVarPrefix("LP"),
-		ff.WithEnvVarSplit(","),
-	)
-	flag.CommandLine.Parse(nil)
-	vFlag.Value.Set(cliFlags.Verbosity)
-
-	err := ValidateFlags(cliFlags)
-	CheckError(err)
+// Run is the entrypoint for main program.
+func Run(buildFlags types.BuildFlags) {
+	cliFlags := cli.GetCliFlags(buildFlags)
 	var waitGroup sync.WaitGroup
 	manifest := ParseYamlManifest(cliFlags.ManifestFile)
 	for _, element := range manifest.Box {
 		if element.Skip {
 			continue
 		}
 		waitGroup.Add(1)
-		DownloadService(cliFlags, manifest, element, &waitGroup)
+		go DownloadService(cliFlags, manifest, element, &waitGroup)
```

Comment:
Is there a reason to pass in `waitGroup` as an argument instead of something like...
```suggestion
		go func(element Service) {
			DownloadService(cliFlags, manifest, element)
			waitGroup.Done()
		}(element)
```

Reply:
ah yes! closures. this is indeed better!